compiler/src/dotty/tools/dotc/interactive/InteractiveDriver.scala

```diff
@@ -224,44 +298,20 @@ class InteractiveDriver(val settings: List[String]) extends Driver {
     new SourceFile(virtualFile, Codec.UTF8)
   }
 
-  def run(uri: URI, sourceCode: String): List[MessageContainer] = run(uri, toSource(uri, sourceCode))
-
-  def run(uri: URI, source: SourceFile): List[MessageContainer] = {
-    val previousCtx = myCtx
-    try {
-      val reporter =
-        new StoreReporter(null) with UniqueMessagePositions with HideNonSensicalMessages
-
-      val run = compiler.newRun(myInitCtx.fresh.setReporter(reporter))
-      myCtx = run.runContext
-
-      implicit val ctx = myCtx
-
-      myOpenedFiles(uri) = source
-
-      run.compileSources(List(source))
-      run.printSummary()
-      val unit = ctx.run.units.head
-      val t = unit.tpdTree
-      cleanup(t)
-      myOpenedTrees(uri) = topLevelClassTrees(t, source)
-      myCompilationUnits(uri) = unit
-
-      reporter.removeBufferedMessages
-    }
-    catch {
-      case ex: FatalError  =>
-        myCtx = previousCtx
-        close(uri)
-        Nil
-    }
+  /**
+   * Initialize this driver and compiler.
+   *
+   * This is necessary because an `InteractiveDriver` can be put to work without having
+   * compiled anything (for instance, resolving a symbol coming from a different compiler in
+   * this compiler). In those cases, an un-initialized compiler may crash (for instance if
+   * late-compilation is needed).
+   */
+  private[this] def initialize(): Unit = {
+    val run = compiler.newRun(myInitCtx.fresh)
+    myCtx = run.runContext
+    run.compileUnits(Nil, myCtx)
```

Comment:
What happens if we don't do compileUnits here ?

Reply:
We crash with a `NullPointerException` in `Run#lateCompile`. `files` will be `null` (`myUnits` will `eq` `myCachedUnits` (both will be `null`))

Reply:
if we just change the default value of `units` to be `Nil` instead of `null`, would that be enough ?

Reply:
Maybe. It looks like `compileUnits` does some more initialisation that I haven't been able to reproduce. With this initialisation code:

```scala
    val run = compiler.newRun(myInitCtx.fresh)
    implicit val ctx = run.runContext
    ctx.initialize()
    myCtx = ctx
```

I get this crash:

```
[info] exception caught when loading class Entity: java.lang.AssertionError: assertion failed: TermRef(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class immutable)),class List),List(TypeRef(ThisType(
TypeRef(NoPrefix,module class model)),class ParamList))),mkString)
[info] exception caught when loading class AlternateConstructors: java.lang.AssertionError: assertion failed: TermRef(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class immutable)),class List),List(Ty
peRef(ThisType(TypeRef(NoPrefix,module class model)),class ParamList))),mkString)
[error] java.lang.AssertionError: assertion failed: TermRef(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class immutable)),class List),List(TypeRef(ThisType(TypeRef(NoPrefix,module class model)),class
 ParamList))),mkString)
[error]         at dotty.DottyPredef$.assertFail(DottyPredef.scala:38)
[error]         at dotty.tools.dotc.core.Types$NamedType.setDenot(Types.scala:1890)
[error]         at dotty.tools.dotc.core.Types$NamedType.withDenot(Types.scala:2114)
[error]         at dotty.tools.dotc.core.Types$TermRef$.apply(Types.scala:2277)
[error]         at dotty.tools.dotc.core.Types$NamedType$.apply(Types.scala:2263)
[error]         at dotty.tools.dotc.core.Types$Type.select(Types.scala:1246)
[error]         at dotty.tools.dotc.typer.TypeAssigner.selectionType(TypeAssigner.scala:235)
[error]         at dotty.tools.dotc.typer.TypeAssigner.accessibleSelectionType(TypeAssigner.scala:261)
[error]         at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:291)
[error]         at dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:427)
[error]         at dotty.tools.dotc.typer.Typer.typeSelectOnTerm$1(Typer.scala:434)
[error]         at dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:461)
[error]         at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1865)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1938)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1970)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1982)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2045)
[error]         at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:664)
[error]         at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1892)
[error]         at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1939)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1970)
[error]         at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1982)
[error]         at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2045)
[error]         at dotty.tools.dotc.typer.Typer.caseRest$1(Typer.scala:1053)
[error]         at dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:1060)
```

This is the assertion that fails:
```scala
      if (ctx.isAfterTyper)
        assert(!denot.isOverloaded || ctx.mode.is(Mode.Printing), this)
```

I can't find what makes the difference when I do `compileUnits`.

Reply:
ok, nervermind then